CONTRIBUTING.md

```diff
@@ -0,0 +1,393 @@
+
+### Suggesting Enhancements
+
+Have an idea for a new feature or improvement? Please [open an issue](https://github.com/florisvb/PyNumDiff/issues/new) to discuss it. See the [Proposing Features](#proposing-features) section for details.
+
+### Contributing Code
+
+1. Look for issues labeled `good first issue` if you're new to the project
+2. Fork the repository
+3. Create a branch for your changes
+4. Make your changes following our guidelines
+5. Submit a pull request
+
+## Development Setup
+
+### Prerequisites
+
+- Python 3.7 or higher
+- Git
+- (Optional) A virtual environment manager (venv, conda, etc.)
+
+### Setting Up the Development Environment
+
+1. **Fork the repository** on GitHub
+
+2. **Clone your fork** locally:
+   ```bash
+   git clone https://github.com/YOUR_USERNAME/PyNumDiff.git
+   cd PyNumDiff
+   ```
+
+3. **Add the upstream repository**:
+   ```bash
+   git remote add upstream https://github.com/florisvb/PyNumDiff.git
+   ```
+
+4. **Create a virtual environment** (recommended):
+   ```bash
+   # Using venv
+   python -m venv venv
+   
+   # Activate on Windows
+   venv\Scripts\activate
+   
+   # Activate on macOS/Linux
+   source venv/bin/activate
+   ```
+
+5. **Install the package in development mode**:
+   ```bash
+   pip install -e .
+   ```
+
+6. **Install development dependencies**:
+   ```bash
+   pip install pytest pylint
+   ```
+
+7. **Verify the installation**:
+   ```bash
+   pytest -s
+   ```
+
+### Project Structure
+
+- `pynumdiff/` - Main source code
+- `examples/` - Jupyter notebook examples
+- `docs/` - Documentation source files
+- `.github/workflows/` - GitHub Actions CI configuration
+- `tests/` - Test files (if applicable)
+
+## Code Style Guidelines
+
+### Python Style Guide
+
+PyNumDiff follows [PEP 8](https://www.python.org/dev/peps/pep-0008/) style guidelines. Here are the key points:
+
+- Use 4 spaces for indentation (no tabs)
+- Maximum line length: 79 characters (soft limit) or 99 characters (hard limit)
+- Use meaningful variable and function names
+- Follow naming conventions:
+  - Functions and variables: `snake_case`
+  - Classes: `PascalCase`
+  - Constants: `UPPER_SNAKE_CASE`
+
+### Code Quality
+
+The project uses `pylint` for code quality checks. Before submitting a PR:
+
+1. **Run pylint** on your changes:
+   ```bash
+   pylint pynumdiff/
+   ```
+
+2. **Or use the project's linting script**:
+   ```bash
+   python linting.py
+   ```
+
+3. **Fix any issues** reported by pylint before submitting your PR
+
+### Editor Configuration
+
+The project includes an `.editorconfig` file that ensures consistent formatting. Most modern editors support EditorConfig automatically.
+
+## Testing Guidelines
+
+### Running Tests
+
+PyNumDiff uses `pytest` for testing. To run tests:
+
+```bash
+# Run all tests
+pytest -s
+
+# Run tests with plots (to visualize method results)
+pytest -s --plot
+
+# Run tests with bounds (to print log error bounds)
+pytest -s --bounds
+```
+
+### Writing Tests
```

Comment:
Might be helpful to mention that the `test_diff_methods` broadly tests for correctness/ability to actually differentiate, that `test_utils` contains tests of miscellaneous stuff like simulations and evaluation metrics and whatnot, and `test_optimize` tests the hyperparameter optimization code, which is its own beast.